api/v1alpha1/bundle_types.go

```diff
@@ -40,9 +40,20 @@ const (
 type BundleSpec struct {
 	// ProvisionerClassName sets the name of the provisioner that should reconcile this BundleInstance.
 	ProvisionerClassName string `json:"provisionerClassName"`
+	// Source defines the configuration for the underlying Bundle content.
+	Source BundleSource `json:"source"`
+}
 
+type BundleSource struct {
+	// Type defines the kind of Bundle content being sourced.
+	Type string `json:"type"`
 	// Image is the bundle image that backs the content of this bundle.
-	Image string `json:"image"`
+	Image ImageSource `json:"image,omitempty"`
```

Comment:
Should this be a pointer to an `ImageSource` because we expect exactly one non-`nil` source struct?

```suggestion
	Image *ImageSource `json:"image,omitempty"`
```

Also, do we need to add any of the union type comment markers that are documented in the KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md?

(I'm wondering if controller-tools has any support for these...)

Reply:
I think a pointer here makes sense. 

I played around with the union type markers locally but it didn't look like the codegen was picking them up. Based on the open https://github.com/kubernetes-sigs/controller-tools/issues/461 I guess controller-gen doesn't have support for union semantics. We would have to hand-craft the CRD schema or move the union type validation to a webhook (the proposed bundle webhook would be a good candidate).

Reply:
Yeah I _think_ this needs to be a pointer too - good catch.

Reply:
Dan and I were playing around kubebuilder markers, and whether we can even manually make direct CRD changes (e.g. basically what I had to do with metering for an API that emulates this design), and as expected, the generation tooling stomps on those direct changes. Maybe this can be tackled in the Bundle webhook @akihikokuroda is working on, or internally in the provisioner's bundle controller?

Reply:
> the generation tooling stomps on those direct changes

Yeah, we either need the generator to understand how to add union semantics or we need a way to patch the generated CRD as part of the generation process.

SDK/Kubebuilder projects use kustomize to make it easy to integrate arbitrary patches into the manifest generation process. Is that a reasonable dependency to carry for this purpose?

Reply:
Personally I would be interested in using kustomize more and potentially bringing it in if it helps deliver this functionality